### PR TITLE
Add warning when using openshift

### PR DESCRIPTION
**Pull request by infiniteregrets**
closes https://github.com/metalbear-co/mirrord/issues/1560

### CODE REVIEW

File: mirrord/cli/src/connection.rs [hunk 1]
```diff
@@ -83,6 +83,8 @@ where
         let k8s_api = KubernetesAPI::create(config)
             .await
             .map_err(CliError::KubernetesApiFailed)?;
+
+        k8s_api.detect_openshift().await?;
```
Comment (aviramha) [inline mirrord/cli/src/connection.rs:87]: We shouldn't exit on this error - just log in info level that we failed trying to detect.

File: mirrord/kube/src/api/kubernetes.rs [hunk 2]
```diff
@@ -56,6 +56,18 @@ impl KubernetesAPI {
...
         }
     }
+
+    pub async fn detect_openshift(&self) -> Result<()> {
+        Discovery::new(self.client.clone())
+            .run()
+            .await
+            .map(|discovery| {
+                if discovery.has_group("route.openshift.io") {
+                    eprintln!("mirrord is running on openshift, due to default PSP of openshift, mirrord may not be able to create the agent. Please refer to https://mirrord.dev/docs/overview/faq/");
```
Comment (aviramha) [inline mirrord/kube/src/api/kubernetes.rs:66]: use the task api?

File: mirrord/cli/src/connection.rs (hunk 1 above)
Comment (aviramha) [reply mirrord/cli/src/connection.rs:87]: we should also do it only if we don't use operator.

**Review by aviramha — CHANGES_REQUESTED** [review]

File: mirrord/kube/src/api/kubernetes.rs [hunk 3]
```diff
@@ -57,6 +57,23 @@ impl KubernetesAPI {
...
+    where
+        P: Progress + Send + Sync,
+    {
+        let mut sub_progress = progress.subtask("checking for openshift");
+        let _= Discovery::new(self.client.clone())
+            .run()
+            .await
+            .map(|api_discovery| {
+                if api_discovery.has_group("route.openshift.io") {
+                    sub_progress.success(Some("mirrord is running on openshift, due to default PSP of openshift, mirrord may not be able to create the agent. Please refer to https://mirrord.dev/docs/overview/faq/"));
```
Comment (meowjesty) [inline mirrord/kube/src/api/kubernetes.rs:71]: Is this a `success`? It kinda reads like a warning?

File: mirrord/kube/src/api/kubernetes.rs [hunk 4]
```diff
@@ -57,6 +57,23 @@ impl KubernetesAPI {
...
+                } else {
+                    sub_progress.success(None);
+                }
+        });
```
Comment (meowjesty) [inline mirrord/kube/src/api/kubernetes.rs:75]:
```suggestion
        if let Some(api_discovery) = Discovery::new(self.client.clone()).run().await && api_discovery.has_group("route.openshift.io")  {
                 sub_progress.success(Some("mirrord is running on openshift, due to default PSP of openshift, mirrord may not be able to create the agent. Please refer to https://mirrord.dev/docs/overview/faq/"));
        };
```

Isn't this just an `if let` as the `None` case ends up being just a noop?

File: mirrord/kube/src/api/kubernetes.rs [hunk 5]
```diff
@@ -57,6 +57,23 @@ impl KubernetesAPI {
             target,
         }
     }
+
+    pub async fn detect_openshift<P>(&self, progress: &P) -> ()
```
Comment (meowjesty) [inline mirrord/kube/src/api/kubernetes.rs:61]: Do we need to return the `()`?

Have not looked into the progress stuff, so my bad if this is a requirement.

**Review by meowjesty — CHANGES_REQUESTED** [review]
Just some questions.

File: mirrord/kube/src/api/kubernetes.rs (hunk 3 above)
Comment (infiniteregrets) [reply mirrord/kube/src/api/kubernetes.rs:71]: Oh my bad, i didnt push the updated branch

File: mirrord/kube/src/api/kubernetes.rs (hunk 4 above)
Comment (infiniteregrets) [reply mirrord/kube/src/api/kubernetes.rs:75]: It returns an Option, I am wondering if we shouldnt eat the error when its not found?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

**Review by eyalb181 — APPROVED** [review]

File: mirrord/cli/src/connection.rs [hunk 6]
```diff
@@ -91,6 +91,8 @@ where
             .await
             .map_err(CliError::KubernetesApiFailed)?;
 
+        k8s_api.detect_openshift(progress).await?;
```
Comment (aviramha) [inline mirrord/cli/src/connection.rs:94]: Create a new task for detecting openshift (progress)

1. When openshift is detected, use the warning api to leave a message about the possible issues
2. When openshift isn't detected, end the progress with "openshift not detected"
3. When you fail detecting, (i.e `detect_openshift` returns an err) use warning API to let user know check wasn't done "couldn't determine if OpenShift"

**Review by aviramha — CHANGES_REQUESTED** [review]

**Review by aviramha — APPROVED** [review]

File: mirrord/kube/src/api/kubernetes.rs [hunk 7]
```diff
@@ -57,6 +57,20 @@ impl KubernetesAPI {
...
+    pub async fn detect_openshift<P>(&self, progress: &P) -> Result<()>
+    where
+        P: Progress + Send + Sync,
+    {
+        if Discovery::new(self.client.clone())
+            .run()
+            .await?
+            .has_group("route.openshift.io")
+        {
+            progress.warning("mirrord is running on openshift, due to default PSP of openshift, mirrord may not be able to create the agent. Please refer to https://mirrord.dev/docs/overview/faq/#can-i-use-mirrord-with-openshift");
```
[the remark this answers is not in the document]
Comment (eyalb181) [reply mirrord/kube/src/api/kubernetes.rs:70]: This seems to have been reverted @infiniteregrets

**Review by meowjesty — APPROVED** [review]
:bear: :+1: 

But there are some clippy warnings though.

**Pull request merged**